zgw/notificaties-api/src/main/kotlin/com/ritense/notificatiesapi/NotificatiesApiPlugin.kt

```diff
@@ -16,19 +16,105 @@
 
 package com.ritense.notificatiesapi
 
+import com.ritense.notificatiesapi.client.NotificatiesApiClient
+import com.ritense.notificatiesapi.domain.Abonnement
+import com.ritense.notificatiesapi.domain.Kanaal
+import com.ritense.notificatiesapi.domain.NotificatiesApiAbonnementLink
+import com.ritense.notificatiesapi.domain.NotificatiesApiConfigurationId
+import com.ritense.notificatiesapi.repository.NotificatiesApiAbonnementLinkRepository
 import com.ritense.plugin.annotation.Plugin
 import com.ritense.plugin.annotation.PluginProperty
+import com.ritense.plugin.domain.PluginConfigurationId
 import java.net.URI
+import java.security.SecureRandom
+import java.util.Base64
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.runBlocking
+import mu.KotlinLogging
 
 @Plugin(
     key = "notificatiesapi",
     title = "Notificaties API",
     description = "Enable interfacing with Notificaties API specification compliant APIs"
 )
-class NotificatiesApiPlugin {
+class NotificatiesApiPlugin(
+    private val client: NotificatiesApiClient,
+    private val abonnementLinkRepository: NotificatiesApiAbonnementLinkRepository
+) {
     @PluginProperty(key = "url", secret = false)
     lateinit var url: URI
 
     @PluginProperty(key = "authenticationPluginConfiguration", secret = false)
     lateinit var authenticationPluginConfiguration: NotificatiesApiAuthentication
+
+    fun createAbonnement(
+        pluginConfigurationId: PluginConfigurationId,
+        callbackUrl: String,
+        kanaalNames: Set<String> = DEFAULT_KANALEN_NAMES
+    ) {
+        ensureKanalenExist(kanaalNames)
+        runBlocking {
+            client.createAbonnement(
+                authenticationPluginConfiguration,
+                url,
+                Abonnement(
+                    callbackUrl = callbackUrl,
+                    auth = createRandomKey(),
+                    url = url.toASCIIString(),
+                    kanalen = kanaalNames.map { Abonnement.Kanaal(naam = it) }
+                )
+            )
+        }.let {
+            abonnementLinkRepository.save(
+                NotificatiesApiAbonnementLink(
+                    pluginConfigurationId = pluginConfigurationId,
+                    url = it.url!!,
+                    auth = it.auth
+                )
+            )
+        }
+    }
+
+    fun deleteAbonnement(pluginConfigurationId: PluginConfigurationId) {
+        val pluginId = NotificatiesApiConfigurationId(pluginConfigurationId.id)
+
+        abonnementLinkRepository.findById(pluginId)
+            .ifPresent {
```

Comment:
If an abonnement isn't present when this method is called, shouldn't that at least result in a log message as well?

Reply:
This is the way it was done in the Connector for opennotificaties. If it's not present when `deleteAbonnement` is called then it doesn't matter since the expected outcome of this method is to remove it from the database.

I do however see the need to log this, since it might point towards an issue with the database if an entity that is expected to be there isn't.
I'll add a log entry.